ducktape/tests/runner_client.py

```diff
@@ -0,0 +1,281 @@
+        self.log(logging.INFO, "Setting up...")
+        self.test.setUp()
+
+    def run_test(self):
+        """Run the test!
+
+        We expect test_context.function to be a function or unbound method which takes an
+        instantiated test object as its argument.
+        """
+        self.log(logging.INFO, "Running...")
+        return self.test_context.function(self.test)
+
+    def teardown_test(self, teardown_services=True):
+        """teardown method which stops services, gathers log data, removes persistent state, and releases cluster nodes.
+
+        Catch all exceptions so that every step in the teardown process is tried, but signal that the test runner
+        should stop if a keyboard interrupt is caught.
+        """
+        self.log(logging.INFO, "Tearing down...")
+        exceptions = []
+        if hasattr(self.test_context, 'services'):
+            services = self.test_context.services
+
+            # stop services
+            if teardown_services:
+                try:
+                    services.stop_all()
+                except BaseException as e:
+                    exceptions.append(e)
+                    self.log(logging.WARN, "Error stopping services: %s" % e.message + "\n" + traceback.format_exc(limit=16))
+
+            # always collect service logs
+            try:
+                self.test.copy_service_logs()
+            except BaseException as e:
+                exceptions.append(e)
+                self.log(logging.WARN, "Error copying service logs: %s" % e.message + "\n" + traceback.format_exc(limit=16))
+
+            # clean up stray processes and persistent state
+            if teardown_services:
+                try:
+                    services.clean_all()
+                except BaseException as e:
+                    exceptions.append(e)
+                    self.log(logging.WARN, "Error cleaning services: %s" % e.message + "\n" + traceback.format_exc(limit=16))
+
+        try:
+            self.test.free_nodes()
+        except BaseException as e:
+            exceptions.append(e)
+            self.log(logging.WARN, "Error freeing nodes: %s" % e.message + "\n" + traceback.format_exc(limit=16))
+
+        # Remove reference to services. This is important to prevent potential memory leaks if users write services
+        # which themselves have references to large memory-intensive objects
+        del self.test_context.services
+
+    def log(self, log_level, msg):
+        """Log to the service log and the test log of the current test."""
+
+        if not getattr(self, 'test'):
+            msg = "%s: %s" % (self.__class__.__name__, str(msg))
+            self.logger.log(log_level, msg)
+        else:
+            msg = "%s: %s: %s" % (self.__class__.__name__, self.test_context.test_name, str(msg))
+            self.logger.log(log_level, msg)
+            self.test.logger.log(log_level, msg)
+
+        self.send(self.request.log(msg, level=log_level))
```

Comment:
Oops, that' should've been two places (I'll delete `self.test.logger.log(log_level, msg)` since `self.test.logger` and `self.logger` are the same object)

Two places continues the previous logging behavior, where notable events from the "thing running the test" are propagated both the the session_log and the test_log